### PR TITLE
TFM Activities

### PR DESCRIPTION
**Pull request by abhi-markan**
Add activity to the deal

    1. New activity create function to add deal and facility creation activities to the TFM
    2. Addition of new constants
    3. Updated Package.json
    4. Lint updates

### CODE REVIEW

File: azure-functions/acbs-function/api-dtfs-cental-api.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      },
+    });
+    return response.data.deal;
+  } catch (error) {
+    console.error(`ACBS - Error while fetching deal ${dealId}.`, { error });
+    return false;
+  }
+};
+
+const createActivity = async (dealId, activity) => {
```
Comment (ttbarnes) [inline azure-functions/acbs-function/api-dtfs-cental-api.js:23]: Should we call this `updateTfmDeal` and then pass the`tfm.activities`object in from wherever it's called?

Just then if we need to update a TFM deal from somewhere else in ACBS, can just call this function. ?

Also should we call `getDeal` `getTfmDeal`so it's clearer?

File: azure-functions/acbs-function/api-dtfs-cental-api.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+const axios = require('axios');
```
Comment (IgnatG) [inline azure-functions/acbs-function/api-dtfs-cental-api.js:1]: `api-dtfs-cental-api.js` is this filename correct?

Comment (abhi-markan) [reply azure-functions/acbs-function/api-dtfs-cental-api.js:1]: Yep

File: trade-finance-manager-api/graphql-query-tests/mock-deal.js [hunk 3]
```diff
@@ -238,6 +238,7 @@ const MOCK_DEAL = {
     submissionDetails: {
       supplierPartyUrn: 'MOCK supplierPartyUrn',
     },
+    activities: [],
```
Comment (ttbarnes) [inline trade-finance-manager-api/graphql-query-tests/mock-deal.js:241]: I don't think we need this - looks like the `submissionDetails` is actually invalid as not in the graphQL schema

File: azure-functions/acbs-function/api-dtfs-cental-api.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+const centralAPI = process.env.DTFS_CENTRAL_API;
+
+require('dotenv').config();
```
Comment (IgnatG) [inline azure-functions/acbs-function/api-dtfs-cental-api.js:5]: `process.env` is dependant on `dotenv`. so this line should be before `const centralAPI = ...`

File: trade-finance-manager-api/graphql-query-tests/mutation-update-activity-comment.api-test.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
+const { createTestClient } = require('apollo-server-testing');
+const { ApolloServer } = require('apollo-server-express');
+const { applyMiddleware } = require('graphql-middleware');
+const { makeExecutableSchema } = require('graphql-tools');
```
Comment (IgnatG) [inline trade-finance-manager-api/graphql-query-tests/mutation-update-activity-comment.api-test.js:4]: `graphql-tools` is deprecated. use `@graphql-tools/schema`
i.e. `const { makeExecutableSchema } = require('@graphql-tools/schema');`

File: trade-finance-manager-ui/templates/case/activity/_macros/activity-timeline.njk [hunk 6]
```diff
@@ -0,0 +1,84 @@
+{%- from "moj/components/timeline/macro.njk" import mojTimeline -%}
+{% set tfm = params.tfm %}
+
+
+
+{{ mojTimeline({
```
Comment (IgnatG) [inline trade-finance-manager-ui/templates/case/activity/_macros/activity-timeline.njk:6]: is the content of this file being used anywhere?

File: azure-functions/acbs-function/api-ukef-mdm-ea.js [hunk 7]
```diff
@@ -27,8 +31,11 @@ const getCurrency = (currencyCode) => getAPI(`currency/${currencyCode}`);
 
 const getACBSIndustrySector = async (industryId) => getAPI(`map-industry-sector?ukefIndustryId=${industryId}`);
 
+<<<<<<< HEAD
 const getACBSCountryCode = async (country) => getAPI(`country?searchtext=${country}`);
 
+=======
+>>>>>>> be930c9bc (Add activity to the deal)
```
Comment (ttbarnes) [inline azure-functions/acbs-function/api-ukef-mdm-ea.js:38]: typo 😝

File: azure-functions/acbs-function/helpers/activity.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        activities: result,
+      };
+      api.updateTfmDeal(dealId, tfm);
+    });
+  }
+};
+
+module.exports = {
+  add,
+};
```
Comment (ttbarnes) [inline azure-functions/acbs-function/helpers/activity.js:108]: Are we able to add tests for these - I think there were challenges around adding tests to ACBS but what about unit tests?

File: azure-functions/acbs-function/package.json [hunk 9]
```diff
@@ -15,6 +15,7 @@
     "axios": "^0.21.1",
     "dotenv": "^8.2.0",
     "durable-functions": "^1.5.2",
-    "moment": "^2.29.1"
+    "moment": "^2.29.1",
+    "date-fns": "^2.25.0"
```
Comment (ttbarnes) [inline azure-functions/acbs-function/package.json:19]: doesn't look like is used in acbs-function ?

File: dtfs-central-api/src/v1/controllers/tfm/deal/tfm-update-deal.controller.js [hunk 10]
```diff
@@ -42,6 +47,27 @@ const updateDeal = async (dealId, dealChanges, existingDeal) => {
...
+    };
+
+    if (tfmUpdate.tfm.activities) {
+      dealUpdate.tfm.activities = [
+        ...existingDeal.tfm.activities,
+        tfmUpdate.tfm.activities,
+      ];
+    }
+  }
+
```
Comment (ttbarnes) [inline dtfs-central-api/src/v1/controllers/tfm/deal/tfm-update-deal.controller.js:70]: Need to add a test for this, the endpoint has some tests in tfm-deal-put-tfm.api-test.js

File: trade-finance-manager-api/src/graphql/resolvers/filters/filterActivities.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  if (filterType === FILTER_TYPE.COMMENT) {
+    return filterByComment(activities);
+  }
+
+  return activities;
+};
+
+module.exports = {
+  filterActivities,
+};
```
Comment (ttbarnes) [inline trade-finance-manager-api/src/graphql/resolvers/filters/filterActivities.js:27]: need tests for these functions, there is example with `filterTasks` is same directory

File: trade-finance-manager-ui/server/graphql/queries/deal-query.js [hunk 12]
```diff
@@ -1,8 +1,8 @@
 const gql = require('graphql-tag');
 
 const dealQuery = gql`
-  query Deal($_id: String! $tasksFilters: TasksFilters) {
-    deal(params: { _id: $_id, tasksFilters: $tasksFilters }) {
+  query Deal($_id: String! $tasksFilters: TasksFilters $activityFilters: ActivityFilters) {
+    deal(params: { _id: $_id, tasksFilters: $tasksFilters activityFilters: $activityFilters }) {
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/graphql/queries/deal-query.js:5]: Can we update the API test to have all the new things? in `graphql-query-tests/get-deal.api-test.js`

File: trade-finance-manager-ui/server/routes/case/index.js [hunk 13]
```diff
@@ -37,6 +37,9 @@ router.get('/:_id/parties/bond-beneficiary', partiesController.getBondBeneficiar
 router.post('/:_id/parties/bond-beneficiary', caseController.postTfmFacility);
 
 router.get('/:_id/activity', activityController.getActivity);
+router.post('/:_id/activity', activityController.filterActivities);
+router.get('/:_id/activity/post-comment', activityController.getCommentBox);
+router.post('/:_id/activity/post-comment', activityController.postComment);
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/routes/case/index.js:42]: unit test in the same directory needs the new routes added

File: trade-finance-manager-ui/templates/case/activity/_macros/activity-comment-filter.njk [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+                'onChange': 'this.form.submit()'
+                }
+            },
+            {
+            value: "comments-only",
+            text: "Comments only",
+            checked: selectedActivityFilter === 'comments-only',
+            attributes: {
+                'data-cy': 'activity-comment-radio-button-comments-only',
+                'onChange': 'this.form.submit()'
```
Comment (ttbarnes) [inline trade-finance-manager-ui/templates/case/activity/_macros/activity-comment-filter.njk:36]: need no-JS handling (show submit button if no JS)

File: trade-finance-manager-ui/templates/case/activity/activity.njk [hunk 15]
```diff
@@ -10,21 +14,37 @@
...
+        items: activities
+      }) }}
 
     </div>
-
+    
   </div>
 
-{% endblock %}
+{% endblock %}
```
Comment (ttbarnes) [inline trade-finance-manager-ui/templates/case/activity/activity.njk:50]: Need to add e2e tests for 
- adding comment
- adding comment validation
- pressing cancel on comment submit form goes back to correct place
- comment displays after submit

File: trade-finance-manager-ui/templates/case/activity/activity-comment.njk [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+        classes: "govuk-label--l",
+        isPageHeading: false
+    },
+    hint: {
+        text: ""
+    },
+    attributes: {
+            'data-cy': 'activity-comment'
+        }
+    }) }}
```
Comment (ttbarnes) [inline trade-finance-manager-ui/templates/case/activity/activity-comment.njk:26]: Looks like need to add validation handling annd probably max comment length?

File: azure-functions/acbs-function/package.json (hunk 9 above)
Comment (abhi-markan) [reply azure-functions/acbs-function/package.json:19]: Yep, I use getUnixTime( to get EPOCH from string datetime.

File: azure-functions/acbs-function/helpers/activity.js (hunk 8 above)
Comment (abhi-markan) [reply azure-functions/acbs-function/helpers/activity.js:108]: Test in durable functions, won't work as its hard to trigger them unless sending a deal object.

Comment (ttbarnes) [reply azure-functions/acbs-function/helpers/activity.js:108]: But aren't the helper files pure JS, I was thinking could use jest for these?

Comment (abhi-markan) [reply azure-functions/acbs-function/helpers/activity.js:108]: Oh yeah, true we can use Jest sorry mate I thought you were talking about E2E, will do.

Comment (ttbarnes) [reply azure-functions/acbs-function/helpers/activity.js:108]: No worries E2E we shouldn't do for ACBS I don't think, ACBS technically has nothing to do with the E2E flow. 

Just concerned that we have a lot of helper/mapping functions for ACBS with nothing tested, will be great to make a start on it 👍

File: trade-finance-manager-api/src/graphql/resolvers/filters/filterActivities.js (hunk 11 above)
Comment (Zainzzkk) [reply trade-finance-manager-api/src/graphql/resolvers/filters/filterActivities.js:27]: done

File: trade-finance-manager-ui/server/graphql/queries/deal-query.js (hunk 12 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/graphql/queries/deal-query.js:5]: done

File: trade-finance-manager-ui/server/routes/case/index.js (hunk 13 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/routes/case/index.js:42]: done

File: trade-finance-manager-ui/templates/case/activity/activity.njk (hunk 15 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/templates/case/activity/activity.njk:50]: done

File: trade-finance-manager-ui/templates/case/activity/activity-comment.njk (hunk 16 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/templates/case/activity/activity-comment.njk:26]: done

File: trade-finance-manager-ui/templates/case/activity/_macros/activity-comment-filter.njk (hunk 14 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/templates/case/activity/_macros/activity-comment-filter.njk:36]: done

**Pull request closed without merging**